File: Python/ByTier/E/1838_E_Count_Supersequences.py
# Problem: CF 1838 E - Count Supersequences
# https://codeforces.com/contest/1838/problem/E

"""
Algorithm: Count Supersequences
Techniques: Modular Arithmetic, Combinatorics, Inclusion-Exclusion Principle

Time Complexity: O(n * log(mod)) per test case due to modular exponentiation
Space Complexity: O(1) additional space (excluding input storage)

This solution uses the inclusion-exclusion principle to count the number of arrays b of size m 
with elements in [1,k] that contain array a as a subsequence.

The approach:
1. Calculate total number of possible arrays of size m with elements in [1,k]: k^m
2. Subtract the count of arrays that do NOT contain a as a subsequence
3. For each position in a, we consider how many ways we can place elements such that 
   a is not a subsequence, using inclusion-exclusion
4. Use modular arithmetic for all calculations to avoid overflow
"""

import io
import os
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def pow(x, y, mod):
    # Fast modular exponentiation
    if x == -1:
        return 1 - y % 2 * 2
    ans, u = 1, x
    while y:
        if y & 1:
            ans = ans * u % mod
        y >>= 1
        u = u * u % mod
    return ans


t = int(input())
mod = 10**9 + 7
ans = []
for _ in range(t):
    n, m, k = map(int, input().split())
    a = list(map(int, input().split()))
    u, v, w = 1, 1, pow(k - 1, m, mod)  # u, v are for combinatorics, w is for the inverse
    inv = pow(k - 1, mod - 2, mod)  # Modular inverse of (k-1)
    ng = 0  # count of invalid sequences
    for i in range(n):
        # Add contribution of current element to invalid count using inclusion-exclusion
        ng += u * pow(v, mod - 2, mod) % mod * w % mod
        u, v, w = u * (m - i) % mod, v * (i + 1) % mod, w * inv % mod
    ans0 = (pow(k, m, mod) - ng) % mod  # Total - Invalid = Valid
    ans.append(ans0)
sys.stdout.write("\n".join(map(str, ans)))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1838/1838_E_Count_Supersequences.cpp
// Problem: CF 1838 E - Count Supersequences
// https://codeforces.com/contest/1838/problem/E

/*
 * Problem: Count Supersequences
 * 
 * Purpose: Given an array 'a' of size n and a target array 'b' of size m, 
 *          where all elements are in range [1, k], count how many arrays 'b' 
 *          exist such that 'a' is a subsequence of 'b'. 
 * 
 * Algorithm: 
 *   - Use inclusion-exclusion principle.
 *   - First compute total number of valid arrays of length m: k^m.
 *   - Then subtract the number of arrays where 'a' is NOT a subsequence.
 *   - For each position in 'a', calculate how many ways we can place it 
 *     such that it appears as a subsequence, and apply inclusion-exclusion 
 *     with derangements.
 * 
 * Time Complexity: O(N + M + log(M) + log(k)) per test case, 
 *                  or approximately O(N) for all cases combined due to 
 *                  preprocessing of inverse modulars.
 * Space Complexity: O(N) for storing inverse modulars and input arrays.
 */

#include <cstdio>
#include <iostream>

#define A puts("Yes")
#define B puts("No")
#define fo(i, a, b) for (int(i) = (a); (i) <= (b); (i)++)
#define fd(i, b, a) for (int(i) = (b); (i) >= (a); (i)--)
#define mk(x, y) make_pair((x), (y))
#define lc (o << 1)
#define rc (o << 1 | 1)
using namespace std;
typedef long long ll;
typedef double db;
const ll mo = 1e9 + 7;
const ll inf = 1ll << 30;
const int N = 2e5 + 5;
ll n, a[N], inv[N], m, k, ans, t, z;

// Fast exponentiation to compute a^b mod mo
ll power(ll a, ll b) {
  ll t = 1, y = a % mo;
  while (b) {
    if (b & 1)
      t = t * y % mo;
    y = y * y % mo;
    b /= 2;
  }
  return t;
}

void solve() {
  scanf("%lld %lld %lld", &n, &m, &k);
  fo(i, 1, n) scanf("%lld", &a[i]);
  
  // Precompute inverse of (k-1) modulo mo
  z = power(k - 1, mo - 2);
  
  // Initial value: number of ways to place array 'a' in 'b' such that no element 
  // of 'a' matches exactly at its position (i.e., derangements of 'a').
  // This is initialized to (k-1)^m
  ans = t = power(k - 1, m);
  
  // Apply inclusion-exclusion principle step by step,
  // calculating the number of cases where subsequence is missing
  fo(i, 1, n - 1) {
    // Multiply by binomial coefficient C(m, i) and adjust for derangements
    t = t * (m - i + 1ll) % mo * inv[i] % mo * z % mo;
    ans = (ans + t) % mo;
  }
  
  // Use inclusion-exclusion to get final result:
  // Total = k^m - [bad arrangements where 'a' is not subsequence]
  ans = power(k, m) - ans;
  ans = (ans % mo + mo) % mo;
  printf("%lld\n", ans);
}

int main() {
  int T;
  scanf("%d", &T);
  
  // Precompute modular inverses using the formula inv[i] = -(mo/i) * inv[mo%i] mod mo
  inv[1] = 1;
  fo(i, 2, N - 1) {
    inv[i] = -(mo / i) * inv[mo % i] % mo;
    inv[i] = (inv[i] % mo + mo) % mo;
  }
  
  while (T--) {
    solve();
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/